### ISSUE TITLE
swagger/reader.py expects inline swagger definition (BodyDefinition) to be a dict, but it is a string.

### ISSUE THREAD

**Issue by larkinrobert** [aws/aws-sam-cli#2523]
### Description:

Unable to include inline swagger using BodyDefinition: "swagger" because of this error the function (seems to me) expects a dict but is sent a string. The AWS docs say that it should be a string.  If the swagger is "remote" then (from reading reader.py) a string or a dict is accepted, but if it's inline it must be a dict. 

### Steps to reproduce:

Create a template with a stringified JSON swagger and try to deploy it with "sam deploy"

### Observed result:

File "/opt/pipx/venvs/aws-sam-cli/lib/python3.6/site-packages/samcli/commands/deploy/auth_utils.py", line 128, in _auth_definition_body_and_uri
    for _, verb in swagger.get("paths", {}).items():
AttributeError: 'str' object has no attribute 'get'

### Expected result:

A sam deployment.

### Additional environment details (Ex: Windows, Mac, Amazon Linux etc)

1. OS: Ubuntu 20.04.1 LTS
2. `sam --version`: SAM CLI, version 1.15.0

`Add --debug flag to command you are running`

**Comment by awood45** [comment]
Do you have a reproducing template handy for this error?

**Comment by larkinrobert** [comment]
```Transform AWS::Serverless-2016-10-31 failed with: Invalid Serverless Application Specification document. Number of errors found: 10. Resource with id [correlationIdPatch] is invalid. 'CodeUri' is not a valid S3 Uri of the form 's3://bucket/key' with optional versionId query parameter. Resource with id [proxyPost] is invalid. 'CodeUri' is not a valid S3 Uri of the form 's3://bucket/key' with optional versionId query parameter. Resource with id [tenantConfigGet] is invalid. 'CodeUri' is not a valid S3 Uri of the form 's3://bucket/key' with optional versionId query parameter. Resource with id [tenantConfigPost] is invalid. 'CodeUri' is not a valid S3 Uri of the form 's3://bucket/key' with optional versionId query parameter. Resource with id [tokensDelete] is invalid. 'CodeUri' is not a valid S3 Uri of the form 's3://bucket/key' with optional versionId query parameter. Resource with id [tokensPost] is invalid. 'CodeUri' is not a valid S3 Uri of the form 's3://bucket/key' with optional versionId query parameter. R -----snip----```

**Comment by larkinrobert** [comment]
`  resApiGateway:
    Type: AWS::ApiGateway::RestApi
    Properties:
      Name:
        Fn::Sub: ${StackPrefix}_${StackEnv}_resApiGateway
      StageName: dev
      Body:
        Fn::Sub:
        - '{"openapi":"3.0.1","info":{"title":"testing_local_resApiGateway","version":"1.0"},"paths":{"/tenant/{id}/config":{"post":{"x-amazon-apigateway-integration":{"type":"aws_proxy","uri":"arn:aws:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${TenantConfigPostLambda}/invocations","passthroughBehavior":"when_no_match","httpMethod":"POST"}},"options":{"summary":"CORS
          support","responses":{"200":{"description":"Default response for CORS method","headers":{"Access-Control-Allow-Origin":{"schema":{"type":"string"}},"Access-Control-Allow-Methods":{"schema":{"type":"string"}},"Access-Control-Allow-Headers":{"schema":{"type":"string"}}},"content":{}}},"x-amazon-apigateway-integration":{"type":"mock","responses":{"default":{"statusCode":"200","responseParameters":{"method.response.header.Access-Control-Allow-Methods":"''GET,POST,PUT,PATCH,DELETE''","method.response.header.Access-Control-Allow-Headers":"''Content-Type,X-Amz-Date,Authorization,X-Api-Key,correlation-object,x-amz-security-token''","method.response.header.Access-Control-Allow-Origin":"''*''"},"responseTemplates":{"application/json":"{}\n"}}},"requestTemplates":{"application/json":"{\n  \"statusCode\"
          : 200\n}\n"},"passthroughBehavior":"when_no_match"}}},"/vault":{"post":{"description":"test","x-amazon-apigateway-integration":{"type":"aws_proxy","uri":"arn:aws:apigateway:eu-west-1:lambda:path/2015-03-31/functions/${VaultPostLambda}/invocations","passthroughBehavior":"when_no_match","httpMethod":"POST"}},"options":{"summary":"CORS
          support","responses":{"200":{"description":"Default response for CORS method","headers":{"Access-Control-Allow-Origin":{"schema":{"type":"string"}},"Access-Control-Allow-Methods":{"schema":{"type":"string"}},"Access-Control-Allow-Headers":{"schema":{"type":"string"}}},"content":{}}},"x-amazon-apigateway-integration":{"type":"mock","responses":{"default":{"statusCode":"200","responseParameters":{"method.response.header.Access-Control-Allow-Methods":"''GET,POST,PUT,PATCH,DELETE''","method.response.header.Access-Control-Allow-Headers":"''Content-Type,X-Amz-Date,Authorization,X-Api-Key,correlation-object,x-amz-security-token''","method.response.header.Access-Control-Allow-Origin":"''*''"},"responseTemplates":{"application/json":"{}\n"}}},"requestTemplates":{"application/json":"{\n  \"statusCode\"
          : 200\n}\n"},"passthroughBehavior":"when_no_match"}}},"/dataTest":{"post":{"x-amazon-apigateway-integration":{"type":"aws_proxy","uri":"arn:aws:apigateway:eu-west-1:lambda:path/2015-03-31/functions/arn:aws:lambda:eu-west-1:022462775006:function:testing_local_tokensRetrieveData/invocations","passthroughBehavior":"when_no_match","httpMethod":"POST"}},"options":{"summary":"CORS
          support","responses":{"200":{"description":"Default response for CORS method","headers":{"Access-Control-Allow-Origin":{"schema":{"type":"string"}},"Access-Control-Allow-Methods":{"schema":{"type":"string"}},"Access-Control-Allow-Headers":{"schema":{"type":"string"}}},"content":{}}},"x-amazon-apigateway-integration":{"type":"mock","responses":{"default":{"statusCode":"200","responseParameters":{"method.response.header.Access-Control-Allow-Methods":"''GET,POST,PUT,PATCH,DELETE''","method.response.header.Access-Control-Allow-Headers":"''Content-Type,X-Amz-Date,Authorization,X-Api-Key,correlation-object,x-amz-security-token''","method.response.header.Access-Control-Allow-Origin":"''*''"},"responseTemplates":{"application/json":"{}\n"}}},"requestTemplates":{"application/json":"{\n  \"statusCode\"
          : 200\n}\n"},"passthroughBehavior":"when_no_match"}}},"/proxy":{"post":{"x-amazon-apigateway-integration":{"type":"aws_proxy","uri":"arn:aws:apigateway:eu-west-1:lambda:path/2015-03-31/functions/arn:aws:lambda:eu-west-1:022462775006:function:testing_local_proxyPost/invocations","passthroughBehavior":"when_no_match","httpMethod":"POST"}},"options":{"summary":"CORS
          support","responses":{"200":{"description":"Default response for CORS method","headers":{"Access-Control-Allow-Origin":{"schema":{"type":"string"}},"Access-Control-Allow-Methods":{"schema":{"type":"string"}},"Access-Control-Allow-Headers":{"schema":{"type":"string"}}},"content":{}}},"x-amazon-apigateway-integration":{"type":"mock","responses":{"default":{"statusCode":"200","responseParameters":{"method.response.header.Access-Control-Allow-Methods":"''GET,POST,PUT,PATCH,DELETE''","method.response.header.Access-Control-Allow-Headers":"''Content-Type,X-Amz-Date,Authorization,X-Api-Key,correlation-object,x-amz-security-token''","method.response.header.Access-Control-Allow-Origin":"''*''"},"responseTemplates":{"application/json":"{}\n"}}},"requestTemplates":{"application/json":"{\n  \"statusCode\"
          : 200\n}\n"},"passthroughBehavior":"when_no_match"}}},"/tenant/{id}":{"get":{"x-amazon-apigateway-integration":{"type":"aws_proxy","uri":"arn:aws:apigateway:eu-west-1:lambda:path/2015-03-31/functions/arn:aws:lambda:eu-west-1:022462775006:function:testing_local_tenantConfigGet/invocations","passthroughBehavior":"when_no_match","httpMethod":"POST"}},"options":{"summary":"CORS
          support","responses":{"200":{"description":"Default response for CORS method","headers":{"Access-Control-Allow-Origin":{"schema":{"type":"string"}},"Access-Control-Allow-Methods":{"schema":{"type":"string"}},"Access-Control-Allow-Headers":{"schema":{"type":"string"}}},"content":{}}},"x-amazon-apigateway-integration":{"type":"mock","responses":{"default":{"statusCode":"200","responseParameters":{"method.response.header.Access-Control-Allow-Methods":"''GET,POST,PUT,PATCH,DELETE''","method.response.header.Access-Control-Allow-Headers":"''Content-Type,X-Amz-Date,Authorization,X-Api-Key,correlation-object,x-amz-security-token''","method.response.header.Access-Control-Allow-Origin":"''*''"},"responseTemplates":{"application/json":"{}\n"}}},"requestTemplates":{"application/json":"{\n  \"statusCode\"
          : 200\n}\n"},"passthroughBehavior":"when_no_match"}}},"/tokens":{"post":{"x-amazon-apigateway-integration":{"type":"aws_proxy","uri":"arn:aws:apigateway:eu-west-1:lambda:path/2015-03-31/functions/arn:aws:lambda:eu-west-1:022462775006:function:testing_local_tokensPost/invocations","passthroughBehavior":"when_no_match","httpMethod":"POST"}},"options":{"summary":"CORS
          support","responses":{"200":{"description":"Default response for CORS method","headers":{"Access-Control-Allow-Origin":{"schema":{"type":"string"}},"Access-Control-Allow-Methods":{"schema":{"type":"string"}},"Access-Control-Allow-Headers":{"schema":{"type":"string"}}},"content":{}}},"x-amazon-apigateway-integration":{"type":"mock","responses":{"default":{"statusCode":"200","responseParameters":{"method.response.header.Access-Control-Allow-Methods":"''GET,POST,PUT,PATCH,DELETE''","method.response.header.Access-Control-Allow-Headers":"''Content-Type,X-Amz-Date,Authorization,X-Api-Key,correlation-object,x-amz-security-token''","method.response.header.Access-Control-Allow-Origin":"''*''"},"responseTemplates":{"application/json":"{}\n"}}},"requestTemplates":{"application/json":"{\n  \"statusCode\"
          : 200\n}\n"},"passthroughBehavior":"when_no_match"}}},"/tokens/{id}":{"delete":{"x-amazon-apigateway-integration":{"type":"aws_proxy","uri":"arn:aws:apigateway:eu-west-1:lambda:path/2015-03-31/functions/arn:aws:lambda:eu-west-1:022462775006:function:testing_local_tokensDelete/invocations","passthroughBehavior":"when_no_match","httpMethod":"POST"}},"options":{"summary":"CORS
          support","responses":{"200":{"description":"Default response for CORS method","headers":{"Access-Control-Allow-Origin":{"schema":{"type":"string"}},"Access-Control-Allow-Methods":{"schema":{"type":"string"}},"Access-Control-Allow-Headers":{"schema":{"type":"string"}}},"content":{}}},"x-amazon-apigateway-integration":{"type":"mock","responses":{"default":{"statusCode":"200","responseParameters":{"method.response.header.Access-Control-Allow-Methods":"''GET,POST,PUT,PATCH,DELETE''","method.response.header.Access-Control-Allow-Headers":"''Content-Type,X-Amz-Date,Authorization,X-Api-Key,correlation-object,x-amz-security-token''","method.response.header.Access-Control-Allow-Origin":"''*''"},"responseTemplates":{"application/json":"{}\n"}}},"requestTemplates":{"application/json":"{\n  \"statusCode\"
          : 200\n}\n"},"passthroughBehavior":"when_no_match"}}},"/vault/{vault_id}":{"delete":{"x-amazon-apigateway-integration":{"type":"aws_proxy","uri":"arn:aws:apigateway:eu-west-1:lambda:path/2015-03-31/functions/arn:aws:lambda:eu-west-1:022462775006:function:testing_local_vaultDelete/invocations","passthroughBehavior":"when_no_match","httpMethod":"POST"}},"options":{"summary":"CORS
          support","responses":{"200":{"description":"Default response for CORS method","headers":{"Access-Control-Allow-Origin":{"schema":{"type":"string"}},"Access-Control-Allow-Methods":{"schema":{"type":"string"}},"Access-Control-Allow-Headers":{"schema":{"type":"string"}}},"content":{}}},"x-amazon-apigateway-integration":{"type":"mock","responses":{"default":{"statusCode":"200","responseParameters":{"method.response.header.Access-Control-Allow-Methods":"''GET,POST,PUT,PATCH,DELETE''","method.response.header.Access-Control-Allow-Headers":"''Content-Type,X-Amz-Date,Authorization,X-Api-Key,correlation-object,x-amz-security-token''","method.response.header.Access-Control-Allow-Origin":"''*''"},"responseTemplates":{"application/json":"{}\n"}}},"requestTemplates":{"application/json":"{\n  \"statusCode\"
          : 200\n}\n"},"passthroughBehavior":"when_no_match"}},"patch":{"x-amazon-apigateway-integration":{"type":"aws_proxy","uri":"arn:aws:apigateway:eu-west-1:lambda:path/2015-03-31/functions/arn:aws:lambda:eu-west-1:022462775006:function:testing_local_correlationIdPatch/invocations","passthroughBehavior":"when_no_match","httpMethod":"POST"}}}},"components":{},"x-amazon-apigateway-documentation":{"version":"1.01","createdDate":"2020-12-28T19:53:57Z","documentationParts":[{"location":{"type":"METHOD","path":"/dataTest","method":"OPTIONS"},"properties":{"summary":"CORS
          support"}},{"location":{"type":"METHOD","path":"/proxy","method":"OPTIONS"},"properties":{"summary":"CORS
          support"}},{"location":{"type":"METHOD","path":"/tenant/{id}/config","method":"OPTIONS"},"properties":{"summary":"CORS
          support"}},{"location":{"type":"METHOD","path":"/tenant/{id}","method":"OPTIONS"},"properties":{"summary":"CORS
          support"}},{"location":{"type":"METHOD","path":"/tokens/{id}","method":"OPTIONS"},"properties":{"summary":"CORS
          support"}},{"location":{"type":"METHOD","path":"/tokens","method":"OPTIONS"},"properties":{"summary":"CORS
          support"}},{"location":{"type":"METHOD","path":"/vault/{vault_id}","method":"OPTIONS"},"properties":{"summary":"CORS
          support"}},{"location":{"type":"METHOD","path":"/vault","method":"OPTIONS"},"properties":{"summary":"CORS
          support"}},{"location":{"type":"METHOD","path":"/vault","method":"POST"},"properties":{"description":"test"}},{"location":{"type":"RESPONSE","path":"/dataTest","method":"OPTIONS","statusCode":"200"},"properties":{"description":"Default
          response for CORS method"}},{"location":{"type":"RESPONSE","path":"/proxy","method":"OPTIONS","statusCode":"200"},"properties":{"description":"Default
          response for CORS method"}},{"location":{"type":"RESPONSE","path":"/tenant/{id}/config","method":"OPTIONS","statusCode":"200"},"properties":{"description":"Default
          response for CORS method"}},{"location":{"type":"RESPONSE","path":"/tenant/{id}","method":"OPTIONS","statusCode":"200"},"properties":{"description":"Default
          response for CORS method"}},{"location":{"type":"RESPONSE","path":"/tokens/{id}","method":"OPTIONS","statusCode":"200"},"properties":{"description":"Default
          response for CORS method"}},{"location":{"type":"RESPONSE","path":"/tokens","method":"OPTIONS","statusCode":"200"},"properties":{"description":"Default
          response for CORS method"}},{"location":{"type":"RESPONSE","path":"/vault/{vault_id}","method":"OPTIONS","statusCode":"200"},"properties":{"description":"Default
          response for CORS method"}},{"location":{"type":"RESPONSE","path":"/vault","method":"OPTIONS","statusCode":"200"},"properties":{"description":"Default
          response for CORS method"}}]},"x-amazon-apigateway-gateway-responses":{"UNAUTHORIZED":{"statusCode":401,"responseParameters":{"gatewayresponse.header.Access-Control-Expose-Headers":"''*''","gatewayresponse.header.Access-Control-Allow-Origin":"''*''"},"responseTemplates":{"application/json":"{
          \"message\": $context.error.messageString }"}},"ACCESS_DENIED":{"statusCode":403,"responseParameters":{"gatewayresponse.header.Access-Control-Expose-Headers":"''*''","gatewayresponse.header.Access-Control-Allow-Origin":"''*''"},"responseTemplates":{"application/json":"{
          \"message\": $context.error.messageString }"}}},"x-amazon-apigateway-minimum-compression-size":860}'
        - TenantConfigArn:
            Fn::GetAtt:
            - tenantConfigPost
            - Arn
          VaultPostArn:
            Fn::GetAtt:
            - vaultPost
            - Arn
  resApiGatewayExecutionLogGroup:
    Type: AWS::Logs::LogGroup-------snip`

**Comment by larkinrobert** [comment]
Have tried with BodyDefinition as well.

**Comment by larkinrobert** [comment]
Yeah, I think I'm on a path that may work. Will update here.

**Comment by sriram-mv** [comment]
@larkinrobert do let us know if you were able to solve it, based on any additional information you can give us, we can triage this to ensure this is not a bug.

**Comment by larkinrobert** [comment]
In the end, I was not able to solve this. To declare my swagger inline I was required to to use bodydefinition, but every iteration/variation that I tried led to an error due to not meeting the library's expectations (described above).  The only way I found that allowed to me to use a swagger inline was to use the translated template that results from `sam build` --but I do not want to use sam build in my process...if sam build could be made to simply render the template that would be great.

We are using our own build system and view `sam build` as (at best) a duplication of effort. Before we migrated to swagger we were not using sam build (or package) but successfully running our own build and then deploying the templates we wrote via `sam deploy`. Now, because we want to use an inline swagger definition (not stored in S3) we are forced to run sam build.  The only difference I can discern is that the sam-build rendered template is in yaml and the template I'm starting from is in json.

I am currently focused on work arounds and will not be trying to force the python library to accept the inline swagger.

TLDR; I do not see a path for writing a sam template that includes an inline swagger, in json, and deploying that template directly because the swagger reader will not accept it. I am abandoning that path.
